aspnet/mvc/models/formatting.rst

```diff
@@ -1,8 +1,217 @@
+
+Special Case Formatters
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Some special cases are implemented using built-in formatters, as already described above. By default, ``string`` return types will be formatted as `text/plain`. This behavior can be removed by removing the ``TextPlainFormatter``. Likewise, actions that have a model object return type will return a 204 No Content resposne when returning ``null``. This behavior can be removed by removing the ``HttpNoContentOutputFormatter``. Removing formatters is done in *Startup* when configuring MVC.
+
+Response Format URL Mappings
+----------------------------
+
+Clients can request a particular format as part of the URL, such as in the querystring or part of the path, or by using a format-specific file extension (.xml, .json). The mapping from request path should be specified in the route the API is using. For example:
+
+.. code-block:: c#
+
+  [FormatFilter]
+  public class ProductsController
+  {
+  [Route("[controller]/[action]/[id].{format?}")]
+  public Product GetById(int id)
+  
+This route would allow the requested format to be specified as an optional file extension. The ``[FormatFilter]`` attribute checks for the existence of the format value in the ``RouteData`` and will map the output format to the appropriate formatter when the response is created.
+
+.. list-table:: Examples
+  :header-rows: 1
+  
+  * - Route
+    - Behavior
+  * - ``/products/GetById/5`` 
+    - Would use the default output format
+  * - ``/products/GetById/5.json``
+    - Would use the JSON formatter
+  * - ``/products/GetById/5.xml``
+    - Would use the XML formatter (if configured)
+
+Recommendations
+---------------
+
+There are many options when it comes to returning data from API actions. When creating demos or coding something quickly to see if it will work, simple return types may be appropriate. However, for APIs that will be used in production, you should prefer ``IActionResult`` as your return type for data since it allows you to handle error states and different HTTP status codes explicitly.
+
+.. tip:: Prefer ``IActionResult`` as your action return type, since it can effectively handle error states and different HTTP status codes.
+
+Keep your action methods as small and focused as possible. This helps follow the `Don't Repeat Yourself principle <http://deviq.com/don-t-repeat-yourself/>`, and results in code that has fewer bugs and is easier to maintain. Consider :doc:`/mvc/controllers/filters` as an effective means of pulling cross-cutting concerns and policies out of your APIs.
+
+Built properly with :doc:`/fundamentals/dependency-injection`, you should be able to easily unit test your controllers. However, especially if you're relying on filters to apply certain rules, you will also want to add integration tests to your automated test strategy. Learn more about :doc:`/mvc/controllers/testing`.
```

Comment:
What does this have to do with formatters? Maybe something like this:

Using formatters to apply return type rules can complicate unit testing and will typically require integrations testing.